### PR TITLE
Add scikit-learn

### PR DESCRIPTION
**Pull request by jakirkham**
Adds a recipe I wrote to build `scikit-learn` using our BLAS system.

### CODE REVIEW

File: recipes/scikit-learn/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  url: https://pypi.io/packages/source/s/scikit-learn/scikit-learn-{{ version }}.tar.gz
+  md5: a2f8b877e6d99b1ed737144f5a478dfc
+
+build:
+  number: 200
+  # We lack openblas on Windows, and therefore can't build numpy there either currently.
+  skip: true  # [win]
+  script:
+    - python setup.py install --single-version-externally-managed --record=record.txt
+#    - nosetests sklearn --exe
```
Comment (jakirkham) [inline recipes/scikit-learn/meta.yaml]: I'm not having much luck running the tests. So, if anyone has any suggestions, please let me know.

Comment (jakirkham) [inline recipes/scikit-learn/meta.yaml]: Nevermind. Figured it out. It was getting confused by the installed `sklearn` and the in the `sklearn` in the working directory. Turns out we can run this in the test section, which is more appropriate and solves this problem. So, we have moved it.

File: recipes/scikit-learn/build.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+if [ `uname` == "Darwin" ]
+then
+    export LDFLAGS="-headerpad_max_install_names"
```
Comment (jakirkham) [inline recipes/scikit-learn/build.sh]: Running into another `install_name_tool` issue where the libraries' headers are too small. So, here we try to fix it.

**Comment by jakirkham** [comment]
Due to some very noisy warnings the log length gets a little too long and Travis CI is failing. However, we don't have any real way to control these warnings. Even if we did, we are building a matrix of 6 (3 Python, 2 NumPy) builds and are in the testing stage of the last one. This means we that in a feedstock this is not going to cause a problem as 1 build will easily fit under the Travis CI constraint.

**Comment by jakirkham** [comment]
Feedback welcome in the feedstock.

**Comment by jakirkham** [comment]
@amueller @ogrisel @GaelVaroquaux @larsmans @agramfort @jnothman @TomDLT, just letting you know that we have added `scikit-learn` to [conda-forge]( http://conda-forge.github.io/ ). This will allow users to easily install this with [`conda`]( http://conda.pydata.org/docs/ ) ( [`conda install -c conda-forge scikit-learn`]( http://anaconda.org/conda-forge/scikit-learn ) ). The maintenance burden here is light as all releases are built and deployed automatically with CI. Changes to how it is being built can be controlled by changing the recipe in [this feedstock (repo)]( https://github.com/conda-forge/scikit-learn-feedstock ). If you have any interest in being added as a maintainer (to make new releases for instance), we would be more than happy to add any of you that are interested to the maintainers list.

**Comment by amueller** [comment]
@jakirkham so what is the difference between that install and the standard conda install? This one is using an open recipe? And the idea is to make it possible to make releases independent of continuum?

**Comment by jakirkham** [comment]
Sorry for the long turnaround, @amueller. These are good questions. Let's see if I can answer them. 😄

> what is the difference between that install and the standard conda install? This one is using an open recipe? And the idea is to make it possible to make releases independent of continuum?

There seems to be several entangled questions here. So, let me try to disentangle them.

1. What is technically different about the conda-forge recipe vs. the defaults recipe for scikit-learn?
2. What is the relationship between the channels conda-forge and defaults?
3. Related to the former: what is the relationship between the organizations conda-forge and Continuum?
4. What is conceptually different between the conda-forge and Continuum build models?
5. What are the advantages of using conda-forge?

<br>

Regarding 1, the package at conda-forge uses OpenBLAS; whereas, defaults uses MKL. We currently don't have a Windows build. That would require a Windows build of OpenBLAS, which is under investigation. We also run the full test suite currently before doing any release. It seems Continuum runs selected pieces.

Regarding 2, there has been a lot of discussion about this from the beginning and some discussion about this more recently. See this [comment]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ) for some thoughts on this back in March. Step 1: open sourcing recipes has definitely happened. Not clear on the status of step 2, but suffice it to say that conda-forge is rapidly (or has rapidly) become that community. Step 3 is a bit challenging as a validation process for binaries is tricky. Though there has been push to use the binaries directly before (see this [comment]( https://github.com/conda-forge/boost-feedstock/issues/1#issuecomment-211941935 ) and related discussion). The trend definitely seems to be moving towards using recipes here as source for packages built in `defaults`. Though I couldn't precisely tell you where this is or is not the case. To put it simply, we are moving towards having a standalone distribution at conda-forge. The fact that this is well curated and tested by a large community with input from developers of the software (being packaged) themselves, makes this a highly desirable collection of recipes that Continuum wants to use.

Regarding 3, our community consists of members from Continuum, conda enthusiasts, software developers (interested in simple distribution), thematic communities (e.g. bioconda) that see value in a general purpose community packaging system like conda-forge, etc. This creates a very nice environment as it can be a place to sort out conda issues, software issues, and packaging issues. Having everyone here helps make it easy to sort out these different problems. Keeping close relationships with all these groups is key to making this all work for end users.

Regarding 4, everything here is in the open. All of the recipes are open sourced. Any of the tools we use here are open sourced. This is all done under BSD 3-Clause to ensure the community is accessible to a large audience. Also, the build process proceeds through CIs. As a result, the process is visible and trackable. Thus, the process from a recipe to a final package is open to auditing all the way through. It also greatly simplifies discussions with upstream about problems encountered. Additionally, it allows users to make changes and know that they are being used. However, unlike Continuum we don't do any additional auditing of packages. Once they are built on the CIs successfully, they are shipped. We don't provide any kind of protection whatsoever so the user accepts the risk of using them.

Regarding 5, it seems that many of your questions were focused on this point. Namely what are the advantages of conda-forge? We have a healthy, diverse, open, and well-connected community. Our build process is totally transparent and allows for broad participation. We also have a significant amount of automation that reduces the overhead associated with the build structure and organizational management. All of this simplifies the process of making high quality packages that are broadly available on demand.

<br>
Sorry this is so long. I hope this at least answers some of your questions. Please let us know if you have more.

**Comment by amueller** [comment]
@jakirkham thanks for the comprehensive answer. I kind of missed these recent developments and this is a great summary. This sounds really awesome. I'd recommend posting this reply somewhere, I'm pretty sure others have the same questions ;)

I think it might make sense for me and/or @ogrisel to be added as maintainers for the 0.18 release.

**Comment by jakirkham** [comment]
> @jakirkham thanks for the comprehensive answer.

Of course.

> I kind of missed these recent developments and this is a great summary. This sounds really awesome.

Glad you like it. 😄

> I'd recommend posting this reply somewhere, I'm pretty sure others have the same questions ;)

Yeah, I wanted to practice writing it a few times first. 😆

I think I'll work on FAQ for the [docs]( http://conda-forge.readthedocs.io/ ) or maybe the [webpage]( http://conda-forge.github.io/ ). There could  also be a link from one to the other. Thoughts?

> I think it might make sense for me and/or @ogrisel to be added as maintainers for the 0.18 release.

Excellent. This PR ( https://github.com/conda-forge/scikit-learn-feedstock/pull/1 ) will add both you ( @amueller ) and @ogrisel to the `scikit-learn` feedstock. This will give both of you the ability to make changes as needed. A more detailed description of how that works is in that PR.

**Pull request merged**